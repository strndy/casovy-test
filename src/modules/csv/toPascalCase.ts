export const toPascalCase = (str: string): string => {
    return str
        .replace(/[^a-zA-Z0-9]+(.)/g, (_, chr) => chr.toUpperCase())
        .replace(/[^a-zA-Z0-9]/g, '')
        .replace(/^./, chr => chr.toUpperCase());
};
