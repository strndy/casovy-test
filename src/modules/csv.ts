import { parse } from 'csv-parse';
import * as fs from 'fs';
import { StockEvent } from "../types";
import { toPascalCase } from "../toPascalCase";


export const readCSV = async (filePath: string): Promise<StockEvent[]> => {
    const records: any[] = [];
    const parser = fs
        .createReadStream(filePath)
        .pipe(parse({
            columns: (headers: string[]) => headers.map(toPascalCase),
            skip_empty_lines: true,
            autoParse: true
        }));

    for await (const record of parser) {
        records.push(record);
    }

    return records;
}