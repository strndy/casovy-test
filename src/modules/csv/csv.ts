import { parse } from 'csv-parse';
import * as fs from 'fs';
import { Event } from "../../types";
import { toPascalCase } from "./toPascalCase";


export const readCSV = async (filePath: string): Promise<Event[]> => {
    const records: any[] = [];
    const parser = fs
        .createReadStream(filePath)
        .pipe(parse({
            columns: (headers: string[]) => headers.map(toPascalCase),
            skip_empty_lines: true,
            autoParse: true,
            cast: (value, context) => {
                if (context.column === 'Time') {
                    const date = new Date(value);
                    if (date.toString() === 'Invalid Date') {
                        throw new Error(`Invalid date: ${value}`);
                    }
                    return date;
                }
                if (context.column === 'NoOfShares') {
                    return parseInt(value);
                }
                if (context.column === 'PriceShare') {
                    return parseFloat(value);
                }
                return value;
            }
        }));

    for await (const record of parser) {
        records.push(record);
    }


    return records;
}