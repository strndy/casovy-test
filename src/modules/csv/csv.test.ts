import { readCSV } from "./csv";

describe('readCSV', () => {
  it('should read the CSV file', async () => {
    const result = await readCSV('./test/testfile1.csv');
    expect(result).toBeDefined();
    expect(result[0].Time).toBeInstanceOf(Date);
    expect(result[0].Action).toBe('Deposit');
    expect(result[1].Action).toBe('Market buy');
    // checks snapshot
    expect(result).toMatchSnapshot();
  }); 
});
