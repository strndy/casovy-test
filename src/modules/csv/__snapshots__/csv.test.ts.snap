// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readCSV should read the CSV file 1`] = `
[
  {
    "Action": "Deposit",
    "CurrencyFinraFee": "",
    "CurrencyPriceShare": "",
    "CurrencyResult": "",
    "CurrencyStampDutyReserveTax": "",
    "CurrencyTotal": "CZK",
    "CurrencyTransactionFee": "",
    "CurrencyWithholdingTax": "",
    "ExchangeRate": "",
    "FinraFee": "",
    "ID": "83a055dd-9c43-4e8d-a532-98d19ea1ef55",
    "ISIN": "",
    "Name": "",
    "NoOfShares": NaN,
    "Notes": "Transaction ID: 1745755385679756",
    "PriceShare": NaN,
    "Result": "",
    "StampDutyReserveTax": "",
    "Ticker": "",
    "Time": 2019-12-05T08:36:08.000Z,
    "Total": "2000",
    "TransactionFee": "",
    "WithholdingTax": "",
  },
  {
    "Action": "Market buy",
    "CurrencyFinraFee": "",
    "CurrencyPriceShare": "GBP",
    "CurrencyResult": "CZK",
    "CurrencyStampDutyReserveTax": "CZK",
    "CurrencyTotal": "CZK",
    "CurrencyTransactionFee": "",
    "CurrencyWithholdingTax": "",
    "ExchangeRate": "3.30801",
    "FinraFee": "",
    "ID": "EOF127307064",
    "ISIN": "TESTISINAA",
    "Name": "Teststck1",
    "NoOfShares": 3,
    "Notes": "",
    "PriceShare": 5,
    "Result": "",
    "StampDutyReserveTax": "1.86",
    "Ticker": "AA",
    "Time": 2019-12-05T08:42:00.000Z,
    "Total": "374.44",
    "TransactionFee": "",
    "WithholdingTax": "",
  },
  {
    "Action": "Market buy",
    "CurrencyFinraFee": "",
    "CurrencyPriceShare": "GBP",
    "CurrencyResult": "CZK",
    "CurrencyStampDutyReserveTax": "",
    "CurrencyTotal": "CZK",
    "CurrencyTransactionFee": "",
    "CurrencyWithholdingTax": "",
    "ExchangeRate": "0.04342",
    "FinraFee": "",
    "ID": "EOF127347450",
    "ISIN": "TESTISINBB",
    "Name": "Test Stock 2",
    "NoOfShares": 2,
    "Notes": "",
    "PriceShare": 5,
    "Result": "",
    "StampDutyReserveTax": "",
    "Ticker": "BB",
    "Time": 2020-07-05T13:30:00.000Z,
    "Total": "6904.25",
    "TransactionFee": "",
    "WithholdingTax": "",
  },
  {
    "Action": "Deposit",
    "CurrencyFinraFee": "",
    "CurrencyPriceShare": "",
    "CurrencyResult": "",
    "CurrencyStampDutyReserveTax": "",
    "CurrencyTotal": "CZK",
    "CurrencyTransactionFee": "",
    "CurrencyWithholdingTax": "",
    "ExchangeRate": "",
    "FinraFee": "",
    "ID": "931c968d-aaee-4ac3-a4ef-3847aab1cd44",
    "ISIN": "",
    "Name": "",
    "NoOfShares": NaN,
    "Notes": "Transaction ID: 1855758085629242",
    "PriceShare": NaN,
    "Result": "",
    "StampDutyReserveTax": "",
    "Ticker": "",
    "Time": 2020-12-08T11:36:03.000Z,
    "Total": "370",
    "TransactionFee": "",
    "WithholdingTax": "",
  },
  {
    "Action": "Stop sell",
    "CurrencyFinraFee": "",
    "CurrencyPriceShare": "GBP",
    "CurrencyResult": "CZK",
    "CurrencyStampDutyReserveTax": "",
    "CurrencyTotal": "CZK",
    "CurrencyTransactionFee": "",
    "CurrencyWithholdingTax": "",
    "ExchangeRate": "3.2917",
    "FinraFee": "",
    "ID": "EOF127497999",
    "ISIN": "TESTISINAA",
    "Name": "Teststck1",
    "NoOfShares": 1,
    "Notes": "",
    "PriceShare": 6,
    "Result": "-7.41",
    "StampDutyReserveTax": "",
    "Ticker": "AA",
    "Time": 2021-12-09T07:00:07.000Z,
    "Total": "365.16",
    "TransactionFee": "",
    "WithholdingTax": "",
  },
  {
    "Action": "Limit buy",
    "CurrencyFinraFee": "",
    "CurrencyPriceShare": "GBP",
    "CurrencyResult": "CZK",
    "CurrencyStampDutyReserveTax": "",
    "CurrencyTotal": "CZK",
    "CurrencyTransactionFee": "",
    "CurrencyWithholdingTax": "",
    "ExchangeRate": "0.03297",
    "FinraFee": "",
    "ID": "EOF127596919",
    "ISIN": "TESTISINAA",
    "Name": "Teststck1",
    "NoOfShares": 1,
    "Notes": "",
    "PriceShare": 4,
    "Result": "",
    "StampDutyReserveTax": "",
    "Ticker": "AA",
    "Time": 2021-12-10T07:44:18.000Z,
    "Total": "1373.81",
    "TransactionFee": "",
    "WithholdingTax": "",
  },
  {
    "Action": "Dividend (Ordinary)",
    "CurrencyFinraFee": "",
    "CurrencyPriceShare": "GBP",
    "CurrencyResult": "",
    "CurrencyStampDutyReserveTax": "",
    "CurrencyTotal": "CZK",
    "CurrencyTransactionFee": "",
    "CurrencyWithholdingTax": "GBP",
    "ExchangeRate": "Not available",
    "FinraFee": "",
    "ID": "",
    "ISIN": "TESTISINBB",
    "Name": "Test Stock 2",
    "NoOfShares": 1,
    "Notes": "",
    "PriceShare": 0.15,
    "Result": "",
    "StampDutyReserveTax": "",
    "Ticker": "BB",
    "Time": 2022-12-31T12:42:06.000Z,
    "Total": "4.54",
    "TransactionFee": "",
    "WithholdingTax": "Not available",
  },
  {
    "Action": "Market buy",
    "CurrencyFinraFee": "",
    "CurrencyPriceShare": "GBP",
    "CurrencyResult": "CZK",
    "CurrencyStampDutyReserveTax": "",
    "CurrencyTotal": "CZK",
    "CurrencyTransactionFee": "",
    "CurrencyWithholdingTax": "",
    "ExchangeRate": "0.03385",
    "FinraFee": "",
    "ID": "EOF130966990",
    "ISIN": "TESTISINBB",
    "Name": "Test Stock 2",
    "NoOfShares": 6,
    "Notes": "",
    "PriceShare": 8,
    "Result": "",
    "StampDutyReserveTax": "",
    "Ticker": "BB",
    "Time": 2023-01-13T07:00:06.000Z,
    "Total": "8474.46",
    "TransactionFee": "",
    "WithholdingTax": "",
  },
  {
    "Action": "Market sell",
    "CurrencyFinraFee": "",
    "CurrencyPriceShare": "GBP",
    "CurrencyResult": "CZK",
    "CurrencyStampDutyReserveTax": "",
    "CurrencyTotal": "CZK",
    "CurrencyTransactionFee": "",
    "CurrencyWithholdingTax": "",
    "ExchangeRate": "0.03385",
    "FinraFee": "",
    "ID": "EOF130966991",
    "ISIN": "TESTISINAA",
    "Name": "Teststck1",
    "NoOfShares": 1,
    "Notes": "",
    "PriceShare": 8,
    "Result": "",
    "StampDutyReserveTax": "",
    "Ticker": "AA",
    "Time": 2023-01-13T07:00:06.000Z,
    "Total": "8474.46",
    "TransactionFee": "",
    "WithholdingTax": "",
  },
]
`;
